### PR TITLE
cItemGrid: Allocate storage lazily

### PR DESCRIPTION
**Pull request by peterbell10**
Fixes the high memory use in #4079.

After doing `//sphere chest 20` this saves ~1GB(!) of memory on my machine.

### CODE REVIEW

File: src/ItemGrid.cpp [hunk 1]
```diff
@@ -237,9 +240,15 @@ bool cItemGrid::IsSlotEmpty(int a_X, int a_Y) const
...
+	if (IsCompletelyEmpty())
+	{
+		return;  // Already clear
+	}
+
+	auto Slots = GetSlots();
 	for (int i = 0; i < m_NumSlots; i++)
 	{
-		m_Slots[i].Empty();
+		Slots[i].Empty();
```
Comment (bibo38) [inline src/ItemGrid.cpp]: Why not just free the allocated space of `unique_ptr`? This would also remove the if-condition.

File: src/ItemGrid.cpp [hunk 2]
```diff
@@ -651,13 +711,19 @@ int cItemGrid::GetNextUsedSlot(int a_StartFrom) const
 
 void cItemGrid::CopyToItems(cItems & a_Items) const
 {
+	if (IsCompletelyEmpty())
+	{
+		return;
+	}
```
[the remark this answers is not in the document]
Comment (peterbell10) [reply src/ItemGrid.cpp]: First point is probably just the bad naming.  As for the second point, at first I was thinking about a `lazy_ptr` but I couldn't think of a nice way to call the constructors.  Perhaps a `lazy_array` would be feasible though.

File: src/ItemGrid.cpp [hunk 3]
```diff
@@ -791,3 +863,56 @@ void cItemGrid::TriggerListeners(int a_SlotNum)
...
+bool cItemGrid::IsCompletelyEmpty() const
+{
+	return (m_Slots == nullptr);
+}
+
+
+
+
+
+const cItem & cItemGrid::UncheckedGetSlot(int a_SlotNum) const
```
[the remark this answers is not in the document]
Comment (peterbell10) [reply src/ItemGrid.cpp]: This is unchecked in comparison to `GetSlot` which does bounds checking.

File: src/ItemGrid.cpp [hunk 4]
```diff
@@ -791,3 +863,56 @@ void cItemGrid::TriggerListeners(int a_SlotNum)
...
+	}
+}
+
+
+
+
+
+cItem * cItemGrid::GetSlots()
+{
+	if (m_Slots == nullptr)
```
[the remark this answers is not in the document]
Comment (peterbell10) [reply src/ItemGrid.cpp]: I thought about the `const_cast` trick but the function is so short it didn't seem worth the ugliness, maybe I should read Scott Meyers to see why I'm wrong :)

>a `static_assert` would be useful to check, that `m_Slots` isn't a constant

Not sure what you mean here.

File: src/ItemGrid.cpp [hunk 5]
```diff
@@ -791,3 +863,56 @@ void cItemGrid::TriggerListeners(int a_SlotNum)
...
+		m_Slots.reset(new cItem[m_NumSlots]);
+	}
+	return m_Slots.get();
+}
+
+
+
+
+
+const cItem * cItemGrid::GetSlots() const
```
[the remark this answers is not in the document]
Comment (peterbell10) [reply src/ItemGrid.cpp]: I would never return a `std::unique_ptr<T>&` here (maybe even ever) because there is no implied ownership.  Something like an `array_view` would probably be better but a raw pointer is still appropiate.

`GetSlot` would be a poor name considering there's already a function with that name with a different return type.  Anyway, the main reason for `GetSlots` is avoiding duplicated checks for `m_Slots` being `nullptr`.

File: src/ItemGrid.h [hunk 6]
```diff
@@ -186,12 +184,25 @@ class cItemGrid
...
 	bool             m_IsInTriggerListeners;  ///< Set to true while TriggerListeners is running, to detect attempts to manipulate listener list while triggerring
 
+
+	/** Access to the slots. Always use this instead of m_Slots directly.
+	Will cause allocation if IsCompletelyEmpty() was true before the call. */
+	cItem * GetSlots();
+	const cItem * GetSlots() const;
+
+	/** Returns true if no storage has been allocated yet. */
+	bool IsCompletelyEmpty() const;
```
[the remark this answers is not in the document]
Comment (peterbell10) [reply src/ItemGrid.h]: Because it's possible for storage to be allocated but all the individual `cItem`s are empty.  I would describe that as empty but `IsCompletelyEmpty` would be false.

I'm not really happy with the name though. Maybe `IsStorageAllocated` or something would be better?

File: src/ItemGrid.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (peterbell10) [reply src/ItemGrid.cpp]: I've added the `cLazyArray` class and it seems to work quite well.

File: src/LazyArray.h [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+
+	// STL style interface
+
+	const T * cbegin() const { return data(); }
+	T *        begin()       { return data(); }
+	const T *  begin() const { return cbegin(); }
+
+	const T * cend() const { return data() + m_Size; }
+	T *        end()       { return data() + m_Size; }
+	const T *  end() const { return cend(); }
```
Comment (bibo38) [inline src/LazyArray.h]: You've defined typedefs/using for the iterator types. They should be used here. Then I don't need to change it on 7 places.

File: src/LazyArray.h [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+		"cLazyArray<T>: T must be default constructible");
+public:
+	using value_type = T;
+	using pointer = T *;
+	using const_pointer = const T *;
+	using reference = T &;
+	using const_reference = const T &;
+	using size_type = int;
+	using iterator = pointer;
+	using const_iterator = const_pointer;
```
Comment (bibo38) [inline src/LazyArray.h]: Using the pointers works for iterators, but I don't like the side effect of const iterators initializing the data structure.
Maybe this explains, why you only use the for each loop only sometimes.

So I would suggest a seperate class for (at least) the `const_iterator`. This class stores the current position and a pointer to `cLazyArray` and calls the `GetAt`-Method to access the elements.

This should allow the consistent usage of the for each loop. Furthermore it would remove the `IsStorageAllocated` call from most (not all!) access only functions, like `CopyToItemStack`.

File: src/ItemGrid.cpp (hunk 1 above)
Comment (bibo38) [reply src/ItemGrid.cpp]: My thought on this was, that if the storage should become empty, then we want to deallocate the space again. E.g. If I copy a bunch of empty chests over a bunch of full chests, their memory should be freed.

File: src/ItemGrid.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (bibo38) [reply src/ItemGrid.cpp]: 👍

File: src/ItemGrid.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (bibo38) [reply src/ItemGrid.cpp]: Ok, didn't see the `GetSlot` functions in the changed files 🤦‍

File: src/ItemGrid.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (bibo38) [reply src/ItemGrid.cpp]: Just a side question: Why is there no implied ownership? Because we don't change ownership with the reference. And as long, as nobody has the great idea of saving the reference in a member variable 👿 , we should be fine.
And we get the benefits, that no broken code could ever think, he needs to destroy this pointer: `std::unique_ptr<> my_slots(GetSlots())`

File: src/ItemGrid.h (hunk 6 above)
[the remark this answers is not in the document]
Comment (bibo38) [reply src/ItemGrid.h]: Jep, sounds way better than the old one! The IsEmpty came from my idea, that if the container is empty, we could deallocate the space.

File: src/LazyArray.h [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+		{
+			m_Array.reset(new T[m_Size]);
+		}
+		return m_Array.get();
+	}
+
+	T * data()
+	{
+		auto const_this = const_cast<const cLazyArray *>(this);
+		return const_cast<T *>(const_this->data());
```
[the remark this answers is not in the document]
Comment (peterbell10) [reply src/LazyArray.h]: That makes sense but what if I had `cLazyArray<const int>` (*): `m_Array::element_type` would be `const int` but the `const_cast` would still be okay as it wouldn't actually remove the constness.

(*) Which admittedly is a pretty silly thing to do.

File: src/LazyArray.h [hunk 10]
```diff
@@ -0,0 +1,127 @@
...
+	using pointer = T *;
+	using const_pointer = const T *;
+	using reference = T &;
+	using const_reference = const T &;
+	using size_type = int;
+	using iterator = pointer;
+	using const_iterator = const_pointer;
+	
+	cLazyArray(size_type a_Size) NOEXCEPT:
+		m_Size{ a_Size }
```
[the remark this answers is not in the document]
Comment (peterbell10) [reply src/LazyArray.h]: You can't write narrowing conversions.

File: src/LazyArray.h (hunk 7 above)
Comment (peterbell10) [reply src/LazyArray.h]: Fixed

File: src/ItemGrid.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (peterbell10) [reply src/ItemGrid.cpp]: `GetSlots` was simply meant to observe the slots, just like `data()` currently. If instead you return `std::unique_ptr<T>&` then the caller could "steal" ownership:
```cpp
auto StolenSlots = std::move(Array.GetSlots());
```
I don't want the return value of `GetSlots` to imply that taking ownership is a valid option so I don't think `std::unique_ptr<T>&` is appropriate.

As a general philosophy, `T *` shouldn't be used to transfer ownership in modern code anyway.  In this slightly idealised frame of mind it's obvious that you shouldn't delete the pointer.  At least those are my thoughts.

File: src/ItemGrid.cpp (hunk 1 above)
Comment (peterbell10) [reply src/ItemGrid.cpp]: It's a balance between using more memory and wasting time allocating and deallocating repeatedly.
For example you might have a chest that clears and then fills with new loot when opened.  In which case the deallocation was just a waste of time.

File: src/LazyArray.h [hunk 11]
```diff
@@ -0,0 +1,127 @@
...
+
+	cLazyArray(cLazyArray && a_Other) NOEXCEPT:
+		m_Array{ std::move(a_Other.m_Array) },
+		m_Size{ a_Other.m_Size }
+	{
+	}
+
+	cLazyArray & operator = (const cLazyArray & a_Other)
+	{
+		cLazyArray(a_Other).swap(*this);
```
[the remark this answers is not in the document]
Comment (bibo38) [reply src/LazyArray.h]: Not exactly. I've just looked at the SO explanation of the [Copy-and-Swap-Idiom](https://stackoverflow.com/questions/3279543/what-is-the-copy-and-swap-idiom/3279550#3279550).

TL;DR: The compiler can execute a copy-elision optimisation.

But I don't mind, if you leave it like this.

File: src/ItemGrid.cpp [hunk 12]
```diff
@@ -482,26 +490,28 @@ cItem cItemGrid::RemoveOneItem(int a_SlotNum)
...
 		LOGWARNING("%s: Invalid slot number %d out of %d slots, ignoring the call, returning empty item",
-			__FUNCTION__, a_SlotNum, m_NumSlots
+			__FUNCTION__, a_SlotNum, m_Slots.size()
 		);
 		return cItem();
 	}
 
 	// If the slot is empty, return an empty item
-	if (m_Slots[a_SlotNum].IsEmpty())
+	if (m_Slots.GetAt(a_SlotNum).IsEmpty())
```
[the remark this answers is not in the document]
Comment (bibo38) [reply src/ItemGrid.cpp]: > m_Slots.data()[x] is the same except saving the result of data() can be used to avoid duplicating nullptr checks.

Sorry, but either prove me this optimisation with a profiler or just make the code readable! Our processors all know about [branch prediction](https://en.wikipedia.org/wiki/Branch_predictor) and this is a good example, since we either have a lot of "not taken", when we only access data or we have a lot of "taken", when the array is initialized.
Furthermore I also have a technique to eliminate the nullptr checks:
```cpp
static std::array<MAX_CONTAINER_SIZE, Item> emptyContainer;
class Lazy {
Item *first = &emptyContainer[0]
const operator[](int i) { return first[í]; } // NO nullptr checks necessary!!!
```

Sorry, but I remember myself thinking about optimizing a A*-Algorithm. I first thought about using pointers instead of the real objects in the `priority_queue`(to make swapping faster). But after I've looked at the system monitor, I've noticed, that the program ate the whole RAM and the system moved parts of the program data on the swap. When I realized this, I knew that my optimisation woudn't provide any benefits...

> In non-const contexts, m_Slots[x] will cause allocations because the reference could be used to mutate the container.

Jep, I know. But `IsEmpty` should be const, so m_Slots[x] shouldn't cause an allocation. The whole point of the `cLazyArray` is to hide the fact of lazy storage, where we can. Only in case of things like `m_Slots[x].DamageItem`, I would handle the allocation issues via `IsStorageAllocated`. But most functions only access/search in the array for an item, so I would only have a **few** places of `IsStorageAllocated` 

> I admit it's a bit of a funny dance. Maybe I need to expand on the doxy comments for cLazyArray to clarify what might allocate?

No, I think comments are not the best thing. I would only comment the **few** fast bailouts via `IsStorageAllocated`. Either use the current design of `operator[]` and abstract the laizy initialization away from the user (which **might** cause some unintended initialisations). Or make `operator[]` always return `const cItem` and make an additional function `Change(int)`, which will allocate the storage. Then you will be explicit, by saying `m_Items.Change(x).DamageItem` and `m_Items[x].IsEmpty()`.

File: src/LazyArray.h (hunk 9 above)
[the remark this answers is not in the document]
Comment (bibo38) [reply src/LazyArray.h]: `cLazyArray<const int>` is not the problem, since this would invalidate the template at `const T & operator[]` (double const). I meant the problem if somebody should change `std::unique_ptr<T>` to `std::unique_ptr<const T>` (*), because he figured out, that we only access it in const functions.

(*) Your * and the unique_ptr decreases the chance of anybody doing so. But better make the code DAU secure, than searching hours for the error 😼 .

File: src/LazyArray.h (hunk 8 above)
Comment (bibo38) [reply src/LazyArray.h]: My main point is the consistency. I don't want to see other people refactoring the `for(int i)` loop to an for each loop and causing implicit memory allocation (thus break the goals of refactoring). And commenting every `for(int i)` is just bad.

So either implement the [iterator](http://en.cppreference.com/w/cpp/concept/Iterator) (I don't need a random access iterator, just something with an `operator++` and an `operator*`). I would also do the work of implementing it, if you don't want to implement it 🎢 .

Or if you don't want to go too far in this PR, just drop the whole iterator concept out of `cLazyArray`! Then it can be later implemented in a seperate PR. Furthermore this would also give the benefits of no need to change `CopyToItemStack` and reduce the usage of `IsStorageAllocated`.

File: src/LazyArray.h (hunk 11 above)
[the remark this answers is not in the document]
Comment (peterbell10) [reply src/LazyArray.h]: I think a move should be just as cheap as a swap and so this should be just as efficient*.  It's also worth mentioning that the move version can provide the `noexcept` guarantee making it preferable to an elided copy-assign.

For what it's worth, [C++ core guideline C.60][1] takes the same stance.

\* I'm assuming there aren't any cases where copy-elision is valid but not move. Is this wrong?

[1]:https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#c60-make-copy-assignment-non-virtual-take-the-parameter-by-const-and-return-by-non-const

[the remark this answers is not in the document]
Comment (bibo38) [reply src/LazyArray.h]: > I think a move should be just as cheap as a swap

The move constructor just calls swap, so I don't see the difference.

> It's also worth mentioning that the move version can provide the noexcept guarantee making it preferable to an elided copy-assign

Sorry I don't understand, why my Copy-and-Swap version does not have a `noexcept` gurantee 😕 ? My version just delegates the move and copy version of `operator=` to the Copy or Move constructor by not specifying, how I want the value. So I only need one `operator=` implementation to handle Copy and Move.

P.S.: The main source for the copy-elusion "benefit" is the [Want Speed? Pass by value.](https://web.archive.org/web/20140113221447/http://cpp-next.com/archive/2009/08/want-speed-pass-by-value/) article.

File: src/LazyArray.h (hunk 8 above)
Comment (bibo38) [reply src/LazyArray.h]: Thanks for clarification! Just looked at all the places and the diff has just hidden alle the usages of the index.

File: src/ItemGrid.cpp (hunk 12 above)
[the remark this answers is not in the document]
Comment (bibo38) [reply src/ItemGrid.cpp]: > operator[] has the same behaves as any other container

I think, you mean the signature. Because it has two different behaviours, based on the constness.

> I'm not convinced that having both operator[] and Change be special cases is any clearer.

The question is: Should it be clear to the reader, when the container will allocate space and when not?

If it should be clear, then make `operator[]` consistent (either always allocate space or never allocate space) and implement a second function for the other behaviour (`GetAt` or `Change`). Then the reader always know, that if for example GetAt is called, no storage will be allocated and every other `m_Item[x]` will allocate space.

If the user shouldn't need to care, if the class allocates space or not, then always use `operator[]` and maybe leave the `IsStorageAllocated` bail out at areas, where it would allocate storage, but you can do without it. But then you should add a comment, so that the user knows, that the bail out is used to save an allocation.

[the remark this answers is not in the document]
Comment (peterbell10) [reply src/ItemGrid.cpp]: >use the specialisation GetAt all the time, when you doesn't want to allocate space

That is already the intended use, the fact that `operator[] const` doesn't need to allocate is a happy accident.
Look at `cItemGrid::ChangeSlotCount` for example, at first it uses `GetAt` but after `IsEmpty` is false, you know the array has already allocated and there's no need to avoid using `[]` to modify the array.

To add allocation to `operator [] const`, I think I would need a more compelling argument than just being identical to the non-`const` version.

File: src/LazyArray.h (hunk 11 above)
[the remark this answers is not in the document]
Comment (peterbell10) [reply src/LazyArray.h]: Sorry, [looks like you're right][1].  I was thinking that marking the swapping `operator=` as `noexcept` would unintentionally make `std::is_nothrow_copy_assignable` true but that's not the case.

Still though, [this other article][2] from the same series suggests the "canonical assignment" is often suboptimal at moving. So overall I'd stick to the c++ core guidelines on this one.

[1]:https://wandbox.org/permlink/IL9nt5kPx8Xp59cL
[2]:https://web.archive.org/web/20140111132021/http://cpp-next.com:80/archive/2009/09/your-next-assignment

File: src/ItemGrid.cpp (hunk 12 above)
[the remark this answers is not in the document]
Comment (bibo38) [reply src/ItemGrid.cpp]: > first it uses GetAt but after IsEmpty is false, you know the array has already allocated and there's no need to avoid using [] to modify the array.

Sorry, but I don't agree. Code changes/gets refactored. Maybe somebody has the idea to first calculate the resulting count maybe like `std::min(std::max(0, m_Slots[a_SlotNum].m_ItemCount + a_AddToCount), Handler->GetMaxStackSize()))` for some reason. Now he just made the mistake of always allocating the space.

Furthermore it may be confusing to the reader: Does `GetAt` return something different than `operator[]`? Or when should I now use `GetAt`, `operator[]` or `IsStorageAllocated`?

Why not just call `GetAt` at every place, where you are okay with a constant object? This would make it consistent and remove the usage of some unnecessary [IsStorageAllocated calls](https://github.com/cuberite/cuberite/pull/4083/files#diff-3e8894a3ba7bf8753efec6461864b809R409).

> To add allocation to operator [] const, I think I would need a more compelling argument than just being identical to the non-const version.

Here comes the better argument 👊  : Just delete it 🗑 . Then the behaviour is consistent, the `const_cast` trick is not necessary (and we just saved some bytes).

[the remark this answers is not in the document]
Comment (peterbell10) [reply src/ItemGrid.cpp]: >it may be confusing to the reader

I don't really see how constantly alternating between one and the other would make it any less confusing. For example, RemoveItem would have:
```cpp
if (m_Slots.GetAt(i).IsEqual(a_ItemStack))
{
	int NumToRemove = [...];
	m_Slots[i].m_ItemCount -= NumToRemove;
```
I don't think I could tell the difference between `GetAt(i)` and `[i]` without reading the doxy/code.

>remove the usage of some unnecessary IsStorageAllocated calls.

When it's turning an O(n) operation into O(1) I don't really see it as a waste.  I know it's not realistically going to effect performance but it certainly doesn't hurt.

[the remark this answers is not in the document]
Comment (bibo38) [reply src/ItemGrid.cpp]: > I don't think I could tell the difference between GetAt(i) and [i] without reading the doxy/code.

`GetAt` returns a non modifiable object (const), while `[i]` returns a modifiable object. The method signature is enough to show the difference. And you could put in the comments of `[i]`, that `GetAt` should be preferred, when possible.

>  I know it's not realistically going to effect performance but it certainly doesn't hurt.

First of all it hurts readability. I now have to look up one more function and understand, that it is equivalent to the for loop with an empty list. Furthermore this just destroys the abstraction of the `cLazyArray` class, because I now need to look up, if the extra condition is necessary (maybe it prevents accessing uninitialized space).

And the extra point is, that this condition currently changes the behaviour of some functions. Look at the `Clear` function:

```cpp

void cItemGrid::Clear(void)
{
	if (!m_Slots.IsStorageAllocated())
	{
		return;  // Already clear
	}

	for (int i = 0; i < m_Slots.size(); i++)
	{
		m_Slots[i].Empty();
		TriggerListeners(i);
	}
}
```
The new behaviour is, that if no storage is allocated, we don't call any listeners. But on the other hand, if the storage is allocated (**which doesn't imply, that the container contains items**) we call `TriggerListener` for every inventory slot.

For me the increased readability effort and the different behaviour hurt much more, than the 0,0001% performance gain. I always prefer code, that is easy to read over some subtile performance improvements. And if code gets more compilcated due to performance or some overlooked behaviour, it should [be commented](https://en.wikipedia.org/wiki/Lions%27_Commentary_on_UNIX_6th_Edition,_with_Source_Code#%22You_are_not_expected_to_understand_this%22).

P.S.: Would you change the `find`-Method of the STL-vector to something like this (It finds elements at the end with O(1) 😈 )?:

```cpp
Iterator find(const Elem & toFind) const
{
    if(toFind == *this[size() - 1])
       return end() - 1;

    auto it = begin();
    for(; it < end(); ++it)
        if(toFind == *it)
            break;
    return it;
}
```

[the remark this answers is not in the document]
Comment (peterbell10) [reply src/ItemGrid.cpp]: >The method signature is enough to show the difference.

It doesn't tell me that `GetAt` won't cause allocation which is what actually matters.

>this just destroys the abstraction of the `cLazyArray` class

Maybe we disagree on where responsibility lies here. This is what I think:
`cItemGrid` is the one that abstracts away the fact that the storage might not be allocated. `cLazyArray` is about _embracing_ that. You wouldn't ask for `std::shared_ptr` to hide that it's reference counted or `std::list` to hide that it's a doubly-linked list.  It's not an abstraction failure; it's the whole point of the class.


As for `cItemGrid::Clear`: not calling listeners when empty cells are re-emptied is already happening in `cItemGrid::EmptySlot`.  So listeners have to be tolerant to this behaviour already.

Regarding your `find`: it's not really the same because the work to check `IsStorageAllocated` is always being done anyway.  As a counter-point consider this implementation of find:
```cpp
It find(const Elem & toFind) const
{
    auto ret = end();
    for(auto it = begin(); it < end(); ++it)
        if(toFind == *it)
            ret = it;
    return ret;
}
```
Where even after finding the item it continues to loop over the container anyway.  I see the obviously wasted work and it just seems "immoral".

(Ignoring that both implementations of find give the wrong results when multiple elements match)

File: src/ItemGrid.cpp [hunk 13]
```diff
@@ -316,11 +318,16 @@ int cItemGrid::AddItem(cItem & a_ItemStack, bool a_AllowNewStacks, int a_Priorit
 	if ((a_PrioritarySlot != -1) && !IsValidSlotNum(a_PrioritarySlot))
 	{
 		LOGWARNING("%s: Invalid slot number %d out of %d slots",
-			__FUNCTION__, a_PrioritarySlot, m_NumSlots
```
Comment (bearbin) [inline src/ItemGrid.cpp]: Can you fix "prioritary" while you're here?

File: src/ItemGrid.cpp [hunk 14]
```diff
@@ -345,14 +352,14 @@ int cItemGrid::AddItem(cItem & a_ItemStack, bool a_AllowNewStacks, int a_Priorit
 			// All items fit
 			return a_ItemStack.m_ItemCount;
 		}
-	}  // for i - m_Slots[]
+	}  // for i - Slots[]
```
Comment (bearbin) [inline src/ItemGrid.cpp]: What is the point of this comment change? Does this code not still iterate over `m_Slots`?

File: src/ItemGrid.h [hunk 15]
```diff
@@ -33,12 +34,10 @@ class cItemGrid
 
 	cItemGrid(int a_Width, int a_Height);
 
-	~cItemGrid();
-
 	// tolua_begin
 	int GetWidth   (void) const { return m_Width; }
 	int GetHeight  (void) const { return m_Height; }
-	int GetNumSlots(void) const { return m_NumSlots; }
+	int GetNumSlots(void) const { return static_cast<int>(m_Slots.size()); }
```
Comment (bearbin) [inline src/ItemGrid.h]: What does this cast do? size_type is aliased to int, so this would be casting from int to int.

File: src/ItemGrid.cpp (hunk 13 above)
Comment (peterbell10) [reply src/ItemGrid.cpp]: Done

File: src/ItemGrid.cpp (hunk 14 above)
Comment (peterbell10) [reply src/ItemGrid.cpp]: There used to be an intermediate variable that's not there any more.
Fixed.

File: src/ItemGrid.h (hunk 15 above)
Comment (peterbell10) [reply src/ItemGrid.h]: Removed.
